src/Microsoft.AspNet.Mvc.ModelBinding/Binders/ComplexModelDtoModelBinder.cs

```diff
@@ -19,8 +19,10 @@ public sealed class ComplexModelDtoModelBinder : IModelBinder
                 var dto = (ComplexModelDto)bindingContext.Model;
                 foreach (var propertyMetadata in dto.PropertyMetadata)
                 {
-                    var propertyModelName = ModelBindingHelper.CreatePropertyModelName(bindingContext.ModelName,
-                                                                                       propertyMetadata.PropertyName);
+                    var propertyModelName =
+                        ModelBindingHelper.CreatePropertyModelName(
+                            bindingContext.ModelName,
+                            propertyMetadata.BinderModelName ?? propertyMetadata.PropertyName);
```

Comment:
From header is supposed to use the exact string that is used for name (as it is a greedy model binder which looks for the exact header), while FromQuery, FromForm and FromRoute are supposed to change the name of the property they are looking up. Hence the difference. I do not think consistency is the intent here.